rfcs/83-reflect_generic_id.md

```diff
@@ -0,0 +1,168 @@
+> - **RFCs should avoid ambiguity:** Two developers implementing the same RFC should come up with nearly identical implementations.
+> - **RFCs should be "implementable":** Merged RFCs should only depend on features from other merged RFCs and existing Bevy features. It is ok to create multiple dependent RFCs, but they should either be merged at the same time or have a clear merge order that ensures the "implementable" rule is respected.
+
+First, add a trait combining traits needed for a key to be used in a registry:
+
+```rs
+pub trait TypeKey: Copy + Hash + Eq {}
+```
+
+This example is for use with hash maps.
+
+Second, make `TypeRegistry` generic over the key type:
+
+```rs
+pub struct TypeRegistry<K: TypeKey> {
+    registrations: HashMap<K, TypeRegistration>
+}
+
+impl<K: TypeKey> TypeRegistraty<K> {
+    pub fn add_registration(&mut self, key: K, registration: TypeRegistration) {
+        self.registrations.insert(key, registration);
+    }
+
+    pub fn register<T: RegisterType<K>>(&mut self) {
+        T::register_type(self);
+    }
+}
+```
+
+Third, introduce new trait, `RegisterType<K>`:
+
+```rs
+pub trait RegisterType<K: TypeKey>: GetTypeRegistration {
+    fn register_type(registry: &mut TypeRegistry<K>) {
+        let key = Self::get_key();
+        let registration_value = Self::get_type_registration();
+        registry.add_registration(key, registration_value);
+    }
+    fn get_key() -> K;
+}
+```
+
+This trait could be implemented multiple times with different `K` types, such as `TypeId`, `&'static str`, `Uuid` and so on:
+
+```rs
+struct Foo;
+
+impl GetTypeRegistration for Foo {
+    // ...
+}
+
+// Implementation with a `Uuid`, some boilerplate could be moved into a derive macro and an attribute, like a former `TypeUuid` trait
+impl RegisterType<Uuid> for Foo {
+    fn get_key() -> Uuid {
+        uuid!("8d17d298-5132-4739-9a13-96e93e4b5034")
+    }
+}
+```
+
+This trait could also be implemented in a derive macro or a blanket implementation:
+
+```rs
+// Implementation using `core::any::TypeId`
+impl<T> RegisterType<TypeId> for T
+where
+    T: core::any::Any,
+{
+    fn get_key() -> TypeId {
+        TypeId::of::<T>()
+    }
+}
+
+// Using a type name, ideally it would be a newtype to indicate that this is a type name and not just some arbitrary string
+impl<T> RegisterType<&'static str> for T {
+    fn get_key() -> &'static str {
+        core::any::type_name::<T>()
+    }
+}
+```
+
+## Drawbacks
+
+> Why should we *not* do this?
+
+This introduces a concept of multiple type identifications which might have no connection between them. There would be no singular central type registry, but instead a number of registries, each with a different id type and for a different purpose.
```

Comment:
I think this is probably one of the biggest issues with this approach. If we end up going with a generic type identification system, we need a way to interop between them such that we can have a central registry.

If we don't have that, then the onus to use the correct registry/identification falls on the user.